File: content/vector-search/source/includes/bson-bindata-vectors/code-snippets/go/get-convert-embeddings-existing-data.go
package main

import (
	"bytes"
	"context"
	"encoding/json"
	"fmt"
	"io/ioutil"
	"log"
	"net/http"
	"os"

	"go.mongodb.org/mongo-driver/v2/bson"
	"go.mongodb.org/mongo-driver/v2/mongo"
	"go.mongodb.org/mongo-driver/v2/mongo/options"
)

const (
	batchSize       = 96
	dbName          = "sample_airbnb"
	collName        = "listingsAndReviews"
	model           = "voyage-3-large"
	outputDimension = 1024
)

func main() {
	apiKey := os.Getenv("VOYAGE_API_KEY")
	mongodbURI := os.Getenv("MONGODB_URI")

	if apiKey == "" || mongodbURI == "" {
		log.Fatal("Ensure VOYAGE_API_KEY and MONGODB_URI are set.")
	}

	summaries, err := fetchSummariesFromMongoDB(mongodbURI)
	if err != nil {
		log.Fatalf("Error fetching summaries: %v", err)
	}

	for start := 0; start < len(summaries); start += batchSize {
		end := start + batchSize
		if end > len(summaries) {
			end = len(summaries)
		}

		floatEmbeddings, err := fetchEmbeddingsFromVoyage(apiKey, summaries[start:end], "float")
		if err != nil {
			log.Fatalf("Error fetching float embeddings: %v", err)
		}

		int8Embeddings, err := fetchEmbeddingsFromVoyage(apiKey, summaries[start:end], "int8")
		if err != nil {
			log.Fatalf("Error fetching int8 embeddings: %v", err)
		}

		ubinaryEmbeddings, err := fetchEmbeddingsFromVoyage(apiKey, summaries[start:end], "ubinary")
		if err != nil {
			log.Fatalf("Error fetching ubinary embeddings: %v", err)
		}

		documents := convertEmbeddingsToBSON(summaries[start:end], floatEmbeddings, int8Embeddings, ubinaryEmbeddings)

		err = writeJSONToFile("embeddings.json", documents)
		if err != nil {
			log.Fatalf("Error writing embeddings to JSON: %v", err)
		}
	}

	fmt.Println("Embeddings successfully saved to embeddings.json")
}

// Fetch documents with the summary field from the collection
func fetchSummariesFromMongoDB(uri string) ([]string, error) {
	ctx := context.TODO()
	clientOpts := options.Client().ApplyURI(uri)

	client, err := mongo.Connect(clientOpts)
	if err != nil {
		return nil, fmt.Errorf("failed to connect to MongoDB: %w", err)
	}
	defer func() {
		if err := client.Disconnect(ctx); err != nil {
			log.Fatalf("Failed to disconnect MongoDB client: %v", err)
		}
	}()

	collection := client.Database(dbName).Collection(collName)
	filter := bson.M{"summary": bson.M{"$nin": []interface{}{nil, ""}}}

	cursor, err := collection.Find(ctx, filter, options.Find().SetLimit(50))
	if err != nil {
		return nil, fmt.Errorf("error finding documents: %w", err)
	}
	defer cursor.Close(ctx)

	var summaries []string
	for cursor.Next(ctx) {
		var result struct {
			Summary string `bson:"summary"`
		}
		if err := cursor.Decode(&result); err != nil {
			return nil, fmt.Errorf("error decoding document: %w", err)
		}
		if result.Summary != "" {
			summaries = append(summaries, result.Summary)
		}
	}

	if err := cursor.Err(); err != nil {
		return nil, fmt.Errorf("cursor error: %w", err)
	}

	return summaries, nil
}

// Fetch embeddings using Voyage AI REST API
func fetchEmbeddingsFromVoyage(apiKey string, texts []string, outputDType string) ([]map[string]interface{}, error) {
	url := "https://ai.mongodb.com/v1/embeddings"

	requestBody := map[string]interface{}{
		"input":            texts,
		"model":            model,
		"output_dtype":     outputDType,
		"output_dimension": outputDimension,
		"input_type":       "document",
	}

	requestBytes, err := json.Marshal(requestBody)
	if err != nil {
		return nil, fmt.Errorf("failed to marshal request body: %w", err)
	}

	req, err := http.NewRequestWithContext(context.TODO(), "POST", url, bytes.NewBuffer(requestBytes))
	if err != nil {
		return nil, fmt.Errorf("failed to create HTTP request: %w", err)
	}

	req.Header.Set("Authorization", fmt.Sprintf("Bearer %s", apiKey))
	req.Header.Set("Content-Type", "application/json")

	client := &http.Client{}
	resp, err := client.Do(req)
	if err != nil {
		return nil, fmt.Errorf("failed to make API request: %w", err)
	}
	defer resp.Body.Close()

	if resp.StatusCode != http.StatusOK {
		body, _ := ioutil.ReadAll(resp.Body)
		return nil, fmt.Errorf("unexpected status code %d: %s", resp.StatusCode, string(body))
	}

	var response struct {
		Data []map[string]interface{} `json:"data"`
	}
	if err := json.NewDecoder(resp.Body).Decode(&response); err != nil {
		return nil, fmt.Errorf("failed to parse API response: %w", err)
	}

	return response.Data, nil
}

// Convert embeddings to BSON binary vectors
func convertEmbeddingsToBSON(summaries []string, floatEmbeddings []map[string]interface{}, int8Embeddings []map[string]interface{}, ubinaryEmbeddings []map[string]interface{}) []bson.M {
	var documents []bson.M

	for i, summary := range summaries {
		floatEmbedding := convertInterfaceToFloat32(floatEmbeddings[i]["embedding"].([]interface{}))
		int8Embedding := convertInterfaceToInt8(int8Embeddings[i]["embedding"].([]interface{}))
		ubinaryEmbedding := convertInterfaceToBytes(ubinaryEmbeddings[i]["embedding"].([]interface{}))

		floatVector := bson.NewVector(floatEmbedding)
		int8Vector := bson.NewVector(int8Embedding)
		ubinaryVector, err := bson.NewPackedBitVector(ubinaryEmbedding, 0)
		if err != nil {
			log.Fatalf("Error creating PackedBitVector: %v", err)
		}

		document := bson.M{
			"text":               summary,
			"embeddings_float32": floatVector.Binary(),
			"embeddings_int8":    int8Vector.Binary(),
			"embeddings_int1":    ubinaryVector.Binary(),
		}

		documents = append(documents, document)
	}

	return documents
}

// Write JSON file from in-memory BSON documents
func writeJSONToFile(filename string, docs []bson.M) error {
	file, err := os.Create(filename)
	if err != nil {
		return fmt.Errorf("failed to create file: %w", err)
	}
	defer file.Close()

	var jsonDocuments []json.RawMessage
	for _, document := range docs {
		jsonBytes, err := bson.MarshalExtJSON(document, false, false)
		if err != nil {
			log.Fatalf("Error: %v", err)
		}
		jsonDocuments = append(jsonDocuments, jsonBytes)
	}

	jsonData, err := json.MarshalIndent(jsonDocuments, "", "  ")
	if err != nil {
		return fmt.Errorf("failed to marshal JSON: %w", err)
	}

	_, err = file.Write(jsonData)
	if err != nil {
		return fmt.Errorf("failed to write JSON to file: %w", err)
	}

	return nil
}

// Converts slice of interface{} to []float32
func convertInterfaceToFloat32(data []interface{}) []float32 {
	f32s := make([]float32, len(data))
	for i, v := range data {
		f32s[i] = float32(v.(float64))
	}
	return f32s
}

// Converts slice of interface{} to []int8 safely
func convertInterfaceToInt8(data []interface{}) []int8 {
	ints8 := make([]int8, len(data))
	for i, v := range data {
		switch val := v.(type) {
		case float64:
			ints8[i] = int8(val)
		case int:
			ints8[i] = int8(val)
		default:
			log.Fatalf("Unexpected type %T in int8 embedding at index %d", v, i)
		}
	}
	return ints8
}

// Converts slice of interface{} to []byte safely
func convertInterfaceToBytes(data []interface{}) []byte {
	bytes := make([]byte, len(data))
	for i, v := range data {
		switch val := v.(type) {
		case float64:
			bytes[i] = byte(val)
		case int:
			bytes[i] = byte(val)
		default:
			log.Fatalf("Unexpected type %T in ubinary embedding at index %d", v, i)
		}
	}
	return bytes
}

File: content/vector-search/source/includes/bson-bindata-vectors/code-snippets/java/get-convert-embeddings-existing-data.java
import okhttp3.*;

import com.mongodb.client.FindIterable;
import com.mongodb.client.MongoClient;  
import com.mongodb.client.MongoClients;  
import com.mongodb.client.MongoCollection;  
import com.mongodb.client.MongoDatabase;  
import org.bson.Document;  
import org.bson.BinaryVector;  
import org.slf4j.Logger;  
import org.slf4j.LoggerFactory;  
import org.json.JSONArray;  
import org.json.JSONObject;  
  
import java.io.FileOutputStream;  
import java.io.IOException;  
import java.util.*;  
import java.util.concurrent.TimeUnit;  
  
public class GenerateAndConvertEmbeddings {  
    private static final Logger logger = LoggerFactory.getLogger(GenerateAndConvertEmbeddings.class);  
  
    // Configuration settings  
    private static final String VOYAGE_API_URL = "https://ai.mongodb.com/v1/embeddings"; // Voyage AI API URL  
    private static final String VOYAGE_API_KEY = System.getenv("VOYAGE_API_KEY");         // Voyage API key  
    private static final String MONGODB_URI = System.getenv("MONGODB_URI");               // MongoDB connection URI  
  
    // Timeout values for API requests  
    private static final int CONNECTION_TIMEOUT = 30; // Timeout for API requests  
    private static final int READ_TIMEOUT = 60;       // Timeout for API responses  
  
    public static void main(String[] args) {  
        try {  
            List<String> summaries = fetchSummariesFromMongoDB();  
            if (summaries.isEmpty()) {  
                throw new RuntimeException("No summaries retrieved from MongoDB.");  
            }  
  
            Document bsonEmbeddings = fetchEmbeddingsFromVoyage(summaries, VOYAGE_API_KEY);  
            if (bsonEmbeddings == null || bsonEmbeddings.isEmpty()) {  
                throw new RuntimeException("Failed to fetch embeddings.");  
            }  
  
            convertAndSaveEmbeddings(bsonEmbeddings);  
        } catch (Exception e) {  
            logger.error("Unexpected error: {}", e.getMessage(), e);  
        }  
    }  
  
    // Fetch summaries from MongoDB collection
    private static List<String> fetchSummariesFromMongoDB() {  
        List<String> summaries = new ArrayList<>();  
        if (MONGODB_URI == null || MONGODB_URI.isEmpty()) {  
            throw new RuntimeException("MongoDB URI is not set.");  
        }  
        logger.info("Connecting to MongoDB at URI: {}", MONGODB_URI);  
  
        try (MongoClient mongoClient = MongoClients.create(MONGODB_URI)) {  
            String dbName = "sample_airbnb";  
            String collName = "listingsAndReviews";  
            MongoDatabase database = mongoClient.getDatabase(dbName);  
            MongoCollection<Document> collection = database.getCollection(collName);  
  
            // Filter to exclude null or empty summaries  
            Document filter = new Document("summary", new Document("$nin", Arrays.asList(null, "")));  
            FindIterable<Document> documentsCursor = collection.find(filter).limit(50);  
  
            for (Document doc : documentsCursor) {  
                String summary = doc.getString("summary");  
                if (summary != null && !summary.isEmpty()) {  
                    summaries.add(summary);  
                }  
            }  
            logger.info("Retrieved {} summaries from MongoDB.", summaries.size());  
        } catch (Exception e) {  
            logger.error("Error fetching from MongoDB: {}", e.getMessage(), e);  
            throw new RuntimeException("Failed to fetch data from MongoDB", e);  
        }  
        return summaries;  
    }  
  
    // Fetch embeddings from Voyage AI API for the given data input  
    private static Document fetchEmbeddingsFromVoyage(List<String> data, String apiKey) {  
        if (apiKey == null || apiKey.isEmpty()) {  
            throw new RuntimeException("API key is not set.");  
        }  
  
        OkHttpClient client = new OkHttpClient.Builder()  
                .connectTimeout(CONNECTION_TIMEOUT, TimeUnit.SECONDS)  
                .readTimeout(READ_TIMEOUT, TimeUnit.SECONDS)  
                .build();  
  
        List<List<List<Integer>>> embeddingsByOutputType = new ArrayList<>();  
        List<String> outputDtypes = List.of("float", "int8", "ubinary");  
  
        try {  
            for (String dtype : outputDtypes) {  
                String responseBody = sendRequest(client, apiKey, data, dtype);  
                embeddingsByOutputType.add(parseEmbeddings(responseBody, dtype));  
            }  
        } catch (IOException e) {  
            logger.error("Error fetching embeddings: {}", e.getMessage(), e);  
            throw new RuntimeException("Error fetching embeddings from Voyage AI.", e);  
        }  
  
        // Convert embeddings to BSON  
        return convertEmbeddingsToBson(data, embeddingsByOutputType);  
    }  
  
    // Send API request to Voyage AI  
    private static String sendRequest(OkHttpClient client, String apiKey, List<String> inputData, String outputDtype) throws IOException {  
        String requestBody = new JSONObject()  
                .put("input", inputData)  
                .put("model", "voyage-3-large")  
                .put("input_type", "document")  
                .put("output_dtype", outputDtype)  
                .put("output_dimension", 1024)  
                .toString();  
  
        Request request = new Request.Builder()  
                .url(VOYAGE_API_URL)  
                .post(RequestBody.create(requestBody, MediaType.get("application/json")))  
                .addHeader("Authorization", "Bearer " + apiKey)  
                .build();  
  
        try (Response response = client.newCall(request).execute()) {  
            if (!response.isSuccessful()) {  
                throw new IOException("API error: HTTP " + response.code());  
            }  
            return response.body().string();  
        }  
    }  
  
    // Parse embeddings from Voyage AI API response 
    private static List<List<Integer>> parseEmbeddings(String responseBody, String outputDtype) {  
        JSONObject responseJson = new JSONObject(responseBody);  
        JSONArray dataArray = responseJson.optJSONArray("data");  
  
        if (dataArray == null) {  
            throw new RuntimeException("Invalid response format: 'data' field missing.");  
        }  
  
        List<List<Integer>> embeddings = new ArrayList<>();  
        for (int i = 0; i < dataArray.length(); i++) {  
            JSONArray embeddingVector = dataArray.getJSONObject(i).getJSONArray("embedding");  
  
            List<Integer> vector = new ArrayList<>();  
            for (int j = 0; j < embeddingVector.length(); j++) {  
                int value = embeddingVector.getInt(j);  
  
                // Handle binary quantization offset for signed int8 representations  
                if ("binary".equals(outputDtype)) {  
                    value = value - 128; // Offset binary method  
                }  
  
                vector.add(value);  
            }  
            embeddings.add(vector);  
        }  
        return embeddings;  
    }  
  
    // Convert fetched embeddings into BSON format  
    private static Document convertEmbeddingsToBson(List<String> inputData, List<List<List<Integer>>> embeddingsByOutputType) {  
        List<Document> bsonEmbeddings = new ArrayList<>();  
        for (int i = 0; i < inputData.size(); i++) {  
            Document embedding = new Document()  
                    .append("text", inputData.get(i))  
                    .append("embeddings_float32", BinaryVector.floatVector(listToFloatArray(embeddingsByOutputType.get(0).get(i))))  
                    .append("embeddings_int8", BinaryVector.int8Vector(listToByteArray(embeddingsByOutputType.get(1).get(i))))  
                    .append("embeddings_int1", BinaryVector.packedBitVector(listToByteArray(embeddingsByOutputType.get(2).get(i)), (byte) 0));  
            bsonEmbeddings.add(embedding);  
        }  
        return new Document("data", bsonEmbeddings);  
    }  
  
    // Save BSON embeddings to a JSON file 
    private static void convertAndSaveEmbeddings(Document bsonEmbeddings) {  
        try (FileOutputStream fos = new FileOutputStream("embeddings.json")) {  
            fos.write(bsonEmbeddings.toJson().getBytes());  
            logger.info("Embeddings with BSON vectors have been saved to embeddings.json");  
        } catch (IOException e) {  
            logger.error("Error writing embeddings to file: {}", e.getMessage(), e);  
        }  
    }  
  
    private static float[] listToFloatArray(List<Integer> list) {  
        float[] array = new float[list.size()];  
        for (int i = 0; i < list.size(); i++) {  
            array[i] = list.get(i).floatValue();  
        }  
        return array;  
    }  
  
    private static byte[] listToByteArray(List<Integer> list) {  
        byte[] array = new byte[list.size()];  
        for (int i = 0; i < list.size(); i++) {  
            array[i] = list.get(i).byteValue();  
        }  
        return array;  
    }  
}  
